### ISSUE TITLE
Render visitors and projectors for unions 

### ISSUE THREAD

**Issue by Baccata** [disneystreaming/smithy4s#1141]
Visitors / projectors are interesting as they are more easier to discover when using smithy4s generated interfaces as clients, than patmat-ability. 

```smithy
union MyUnion {
   myInt: Integer
   myString: String 
   myOther: Unit 
} 
```

should lead to the following rendered code (conceptually speaking) : 

```scala
sealed trait MyUnion {
   // new "accept" method taking a visitor and applying the relevant method 
   // NB : the method should be implemented in each ADT member as opposed to 
   // a pat-mat at the trait level. 
   def accept[A](visitor: Visitor[A]) : A 
   // new projector methods 
   def myInt : Option[Int] 
   def myString: Option[String] 
   def myOther: Option[Unit] 
} 

object MyUnion {
  // ...
  trait Visitor[A] {
    def myInt(value: Int) : A
    def myString(value: String) : A
    def myOther() : A 
  } 
  object Visitor {
    trait Default[A] extends Visitor[A] {
      def default: A 
      def myInt(value:Int) : A = default
      def myString(value:String) : A = default
      def myOther() : A = default
    } 
  } 
 // ... 
} 
```

**Comment by kubukoz** [comment]
not opposed to this, but I feel like discoverability of something being a union or not is a language / language tooling issue of Scala. This is nice as a band-aid but it shouldn't be necessary.

**Comment by Baccata** [comment]
> This is nice as a band-aid but it shouldn't be necessary. 

The visitor may not be necessary. I wouldn't necessarily oppose guarding them behind an opt-in flag like we've done for optics. I think they offer a nice alternative to pat-mat that shines in particular when dealing with large ADTs, and they also have the quality of hiding the `Case` suffix. 

However, the projectors are extremely useful, in particular for scripting, and I think it's annoying that there's not an OOTB way of doing it in vanilla Scala without some library. 

The work has already been done/merged in https://github.com/disneystreaming/smithy4s/pull/1144. If you feel strongly about visitors, we can keep the issue open until someone adds an opt-in flag (similar to the one for optics) to make Visitors opt-in instead of defaults. However the projectors will remain there.

**Comment by kubukoz** [comment]
> the projectors are extremely useful, in particular for scripting

that's a good thing to mention in the PR description and later in the release notes :)

personally I wish we had a `(a: U).asInstanceOfOpt[T <: U]: Option[T]` in the stdlib, but that'd also require tooling support so that only valid subtypes are suggested. And you'd still have the `Case` suffix.

> If you feel strongly about visitors

I don't

> we can keep the issue open until someone adds an opt-in flag

frankly I dislike these flags more than the complexity of rendering new constructs: if we're already paying the price of having all the codegen code to have these constructs, a feature flag is only increasing the complexity by introducing branching. I'd render optics at all times same as visitors.

**Comment by kubukoz** [comment]
closing as implemented in #1144 then
